src/nnet3/discriminative-training.cc

```diff
@@ -0,0 +1,602 @@
+// limitations under the License.
+
+#include "nnet3/discriminative-training.h"
+#include "lat/lattice-functions.h"
+#include "cudamatrix/cu-matrix.h"
+
+namespace kaldi {
+namespace discriminative {
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats() {
+  std::memset(this, 0, sizeof(*this));
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = 0; 
+}
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(int32 num_pdfs) {
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = num_pdfs; 
+  gradients.Resize(num_pdfs); 
+  output.Resize(num_pdfs);
+}
+
+// Constructor from config structure
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(
+    DiscriminativeTrainingStatsOptions opts) : config(opts) { 
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// Reset statistics
+void DiscriminativeTrainingStats::Reset() {
+  gradients.SetZero();
+  output.SetZero();
+
+  tot_t = 0.0;
+  tot_t_weighted = 0.0;
+  tot_objf = 0.0;
+  tot_num_count = 0.0;
+  tot_den_count = 0.0;
+  tot_num_objf = 0.0;
+}
+
+// Sets the config structure
+void DiscriminativeTrainingStats::SetConfig(
+    const DiscriminativeTrainingStatsOptions &opts) {
+  config = opts;
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// This class is responsible for the forward-backward of the
+// 'supervision' lattices and computation of the objective function
+// and gradients.
+//
+// note: the supervision.weight is ignored by this class, you have to apply
+// it externally.
+class DiscriminativeComputation {
+  typedef Lattice::Arc Arc;
+  typedef Arc::StateId StateId;
+ 
+ public:
+  // Initialize the objcect.  Note: we expect the 'nnet_output' to have the
+  // same number of rows as supervision.num_frames * supervision.num_sequences,
+  // and the same number of columns as tmodel.NumPdfs(); but the
+  // ordering of the rows of 'nnet_output' is not the same as the ordering of
+  // frames in paths in the 'supervision' object (which has all frames of the
+  // 1st sequence first, then the 2nd sequence, and so on).  Instead, the
+  // frames in 'nnet_output' are ordered as: first the first frame of each
+  // sequence, then the second frame of each sequence, and so on. 
+  // This is done to be similar to the setup in 'chain' training 
+  // even though this does not offer any computational advantages here
+  // as in the 'chain' case.
+  DiscriminativeComputation(const DiscriminativeTrainingOptions &opts,
+      const TransitionModel &tmodel,
+      const CuVectorBase<BaseFloat> &log_priors,
+      const DiscriminativeSupervision &supervision,
+      const CuMatrixBase<BaseFloat> &nnet_output,
+      DiscriminativeTrainingStats *stats,
+      BaseFloat *l2_term,
+      CuMatrixBase<BaseFloat> *nnet_output_deriv,
+      CuMatrixBase<BaseFloat> *xent_output_deriv);
```

Comment:
I don't understand why you have a separate derivative matrix called 'xent_output_deriv'.
The reason this existed in the chain setup was that the nnet has two distinct outputs.
In discriminative training, there will presumably just be one output (?)  Or are you also trying to implement the same kind of xent smoothing as we support in the chain models, with a different output?